plugins/SegmentEditor/API.php

```diff
@@ -238,29 +229,42 @@ private function getModel()
      * @param int $idSegment The ID of the stored segment to modify.
      * @param string $name The new name of the segment.
      * @param string $definition The new definition of the segment.
-     * @param bool $idSite If supplied, associates the stored segment with as single site.
+     * @param int|string|null $idSite If supplied, associates the stored segment with as single site.
      * @param bool $autoArchive Whether to automatically archive data with the segment or not.
      * @param bool $enabledAllUsers Whether the stored segment is viewable by all users or just the one that created it.
-     *
-     * @return bool
      */
-    public function update($idSegment, $name, $definition, $idSite = false, $autoArchive = false, $enabledAllUsers = false)
-    {
+    public function update(
+        int $idSegment,
+        string $name,
+        string $definition,
+        $idSite = null,
+        bool $autoArchive = false,
+        bool $enabledAllUsers = false
+    ): void {
         $segment = $this->getSegmentOrFail($idSegment);
         $this->checkUserCanEditOrDeleteSegment($segment);
 
         $idSite = $this->checkIdSite($idSite);
+        $name = Common::sanitizeInputValue($name);
         $this->checkSegmentName($name);
-        $definition      = $this->checkSegmentValue($definition, $idSite);
-        $enabledAllUsers = $this->checkEnabledAllUsers($enabledAllUsers);
+        $definition = Common::sanitizeInputValue($definition);
+        $definition = $this->checkSegmentValue($definition, $idSite);
+
+        // only check param if value is changed
+        // this ensure that a segment from a user with lower permission can still be changed by them
+        // if a superuser updated the segment to be available for all users
+        if ((int) $segment['enable_all_users'] !== (int) $enabledAllUsers) {
+            $enabledAllUsers = $this->checkEnabledAllUsers($enabledAllUsers);
+        }
+
         $autoArchive     = $this->checkAutoArchive($autoArchive, $idSite);
 
         $bind = array(
             'name'               => $name,
             'definition'         => $definition,
-            'enable_all_users'   => $enabledAllUsers,
-            'enable_only_idsite' => $idSite,
-            'auto_archive'       => $autoArchive,
+            'enable_all_users'   => (int) $enabledAllUsers,
+            'enable_only_idsite' => (int) $idSite,
```

Comment:
If a super user has updated a segment to be available for all sites, the original creator will see a wrong error message if he tries to update it:

> You can only edit and delete custom segments that you have created yourself. This segment was created and 'shared with you' by the Super User. To modify this segment, you can first create a new one by clicking on 'Add new segment'. Then you can customize the segment's definition.

Having a segment availalbe for all sites, even those the creating user does not even have access to, might be a case where we should change the segment owner to the super user that updates this flag? Or show a different error message like "segment only editable by super user"?

Reply:
With the added "only super user can edit"-message that should be addressed, anything we need to do around the deletion of a segment?

Moving the check into `checkUserCanEditOrDeleteSegment` could keep all functionality consistent :thinking: